uaclient/clouds/tests/test_aws.py

```diff
@@ -17,6 +20,44 @@ def test_identity_doc_from_aws_url_pkcs7(self, readurl):
         url = "http://169.254.169.254/latest/dynamic/instance-identity/pkcs7"
         assert [mock.call(url)] == readurl.call_args_list
 
+    @pytest.mark.parametrize("caplog_text", [logging.DEBUG], indirect=True)
+    @pytest.mark.parametrize(
+        "fail_count,exception", ((3, None), (4, HTTPError))
+    )
+    @mock.patch(M_PATH + "util.time.sleep")
+    @mock.patch(M_PATH + "util.readurl")
+    def test_retry_backoff_on_failed_identity_doc(
+        self, readurl, sleep, fail_count, exception, caplog_text
+    ):
+        """Retry backoff is attempted before failing to get AWS.identity_doc"""
+
+        def fake_someurlerrors(url):
+            if readurl.call_count <= fail_count:
+                raise HTTPError(
+                    None, 400 + readurl.call_count, None, None, BytesIO()
+                )
+            return "pkcs7WOOT!==", {"header": "stuff"}
+
+        readurl.side_effect = fake_someurlerrors
```

Comment:
For reference, you could achieve roughly the same effect with:

```py
readurl.side_effect = [
    HTTPError(None, 400 + readurl.call_count, None, None, BytesIO())
] * fail_count + [("pkcs7WOOT!==", {"header": "stuff"})]
```

because `side_effect` will iterate over a given list, raising any exception subclasses and returning the value otherwise. In this case, I think what you went with is preferable, but figured I'd share an interesting tidbit. :+1:

Reply:
Smart idea, the problem is that list arithmetic is setting the error code to 400 for every HTTPError created and I wanted something to reprensent increased different error codes per error. We could do this with a bit more verbose list creation, but I figured that's gets almost as verbose as the assigned local function.